src/middleware.rs

```diff
@@ -36,40 +81,128 @@ impl State {
-                        masking
-                    } else {
-                        &self.sdk.masking
-                    };
-
-                    let har = HarBuilder::new(request, response).build(masking);
-                    println!("HAR BUILT: {:#?}", har);
+                    debug!("response received, request_id: {:?}", &request_id);
+
+                    if let Err(error) = self.build_and_send_har(request_id, request, response) {
+                        error!("Failed to send HAR to Speakeasy: {:#?}", error);
+                    }
                 }
             }
             MiddlewareMessage::ControllerMessage(msg) => {
-                self.controller.handle_message(msg);
+                self.controller_state.handle_message(msg);
+            }
+        }
+    }
+
+    fn build_and_send_har(
+        &mut self,
+        request_id: RequestId,
+        request: GenericRequest,
+        response: GenericResponse,
+    ) -> Result<(), Error> {
+        // look for path hint for request, if not look in the request
+        let path_hint = self
+            .controller_state
+            .get_path_hint(&request_id)
+            .or_else(|| request.path_hint.as_ref().map(ToString::to_string))
+            .unwrap_or_else(|| "".to_string());
+
+        // look for a request specific mask if not use global one
+        let masking = self
+            .controller_state
+            .get_masking(&request_id)
+            .unwrap_or_else(|| self.sdk.masking.clone());
+
+        let config = self.sdk.config.clone();
+        let customer_id = self
+            .controller_state
+            .get_customer_id(&request_id)
+            .unwrap_or_default();
+
+        tokio02::task::spawn(async move {
+            let har = HarBuilder::new(request, response).build(&masking);
+            let har_json = serde_json::to_string(&har).expect("har will serialize to json");
+
+            let masking_metadata = if masking.is_empty() {
+                None
+            } else {
+                Some(masking.into())
+            };
+
+            let ingest = IngestRequest {
+                har: har_json,
+                path_hint,
+                api_id: config.api_id,
+                version_id: config.version_id,
+                customer_id,
+                masking_metadata,
+            };
+
+            if let Err(error) = send(ingest, config.api_key).await {
+                error!("Failed to send HAR to Speakeasy: {:#?}", error);
             }
+        });
+
+        Ok(())
+    }
+}
+
+async fn send(request: IngestRequest, api_key: String) -> Result<(), Error> {
+    // NOTE: Using hyper directly as there seems to be a bug with tonic v0.3 throwing
+    // an error from rustls. When making the middleware for actix4 we can hopefully
+    // avoid doing this and just use the client directly from tonic.
+    let insecure_client = hyper::Client::builder().http2_only(true).build_http();
```

Comment:
Is there a reason that both of these get instantiated with every request?

Reply:
Good catch fixed